CARDS-2558: Clinician dashboard

sdumitriu opened the pull request:
To test:
- build
- run with `./start_cards.sh -P prems  --dev -f 'mvn:io.uhndata.cards/cards-clinician-portal/${CARDS_VERSION}/slingosgifeature' -f 'mvn:io.uhndata.cards/cards-locking/${CARDS_VERSION}/slingosgifeature'`
- create visits, make sure the patient chart, visit chart, forms show up correctly
- also test with a trusted user

marta- committed:
CARDS-2569: Allow configuring the menu items to be displayed for a Form

marta- committed:
CARDS-2558 / CARDS-2559: Visit page view for clinicians
Visit page - added link to patient survey

acrowthe committed:
CARDS-2609: Configurable actions on user dashboard and related views
Implement new clinician views

acrowthe committed:
CARDS-2609: Configurable actions on user dashboard and related views
Finish new clinician views

sdumitriu committed:
CARDS-2461: Clinician-oriented subject and form views
Move the clinician portal code in its own module

sdumitriu committed:
CARDS-2461: Clinician-oriented subject and form views

marta- commented on modules/clinician-portal/pom.xml:
```diff
@@ -0,0 +1,58 @@
+    <version>0.9.31-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>cards-clinician-portal</artifactId>
```
I thought we had decided to call it clinician dashboard, not portal

sdumitriu replied:
Right, done.

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Rename clinician-portal -> clinician-dashboard

veronikaslc commented:
Missing some commits from 2609 (by Andrew), local branch with those commits exists.

acrowthe committed:
CARDS-2609: Configurable actions on user dashboard and related views
Fix merge and linting errors

acrowthe committed:
[misc] Update yarn.lock

acrowthe committed:
CARDS-2609: Configurable actions on user dashboard and related views
- Remove ClinicianForms as it got renamed and then readded due to a merge issue
- Add expand button back to clinician dashboard subjects and forms table

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,94 @@
+    </Tooltip>);
+  }
+  if (fetchingLink) {
+    return (<Tooltip title={"Generating survey link..."}>
```
```suggestion
    return (<Tooltip title="Generating survey link...">
```

veronikaslc commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -510,45 +528,54 @@ function Form (props) {
+      }
+      { isDropdnEnabled() &&
+        <>
+          <Tooltip title="More actions" onClick={(event) => setActionsMenu(event.currentTarget)}>
```
`onClick` is not a `Tooltip` prop

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,94 @@
+      </IconButton>
+    </Tooltip>);
+  }
+  return (<Tooltip title={copied ? "Copied" : `Copy patient survey link to clipboard`}>
```
just out of curiocity - why 2 different quotation ways for simple strings in one line?

acrowthe replied:
No reason now, likely legacy from some point in this branches history. Fixing

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Form.jsx:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  return (
+    <DefaultForm key={window.location} actionSwitches={actionSwitches} />
```
Compiler will memoise this key -> potential stale page

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/FormView.jsx:
```diff
@@ -0,0 +1,54 @@
+
+  const columns = [
+    {
+      "key": "@name",
```
basically you do not need to quote keynames like in lines 25-28

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Forms.jsx:
```diff
@@ -0,0 +1,65 @@
+
+  const columns = [
+    {
+      "key": "@name",
```
same as above - no need to quote keynames

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Patient.jsx:
```diff
@@ -0,0 +1,238 @@
+  //
+
+  const displayPatientInfo = () => {
+    let fName = patientData?.first_name;
```
`.trim()`?

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Patient.jsx:
```diff
@@ -0,0 +1,238 @@
+    let lName = patientData?.last_name;
+    let name = [lName, fName].filter(n => n).join(", ");
+
+    let dobAnswer = patientData?.date_of_birth;
```
can `dobAnswer` be undefined or null?

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,94 @@
+    setFetchingLink(true);
+    fetchWithReLogin(globalLoginDisplay, `${visitURL}.token.html`)
+      .then((response) => response.ok ? response.text() : Promise.reject(response))
+      .then((text) => { setSurveyLink(window.location.origin + "/Survey.html?auth_token=" + text.trim()); copy(window.location.origin + "/Survey.html?auth_token=" + text.trim()); })
```
for better code readability I suggest make each f-n call a separate line there

veronikaslc replied:
arguments for `setSurveyLink()` and `copy()` are identical, should be optimised to be one variable

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,94 @@
+  }
+  if (fetchingLink) {
+    return (<Tooltip title={"Generating survey link..."}>
+      <IconButton size={size || "large"}>
```
`size` is already `"large"` by default as per `defaultProps`

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,94 @@
+    </Tooltip>);
+  }
+  return (<Tooltip title={copied ? "Copied" : `Copy patient survey link to clipboard`}>
+    <IconButton size={size || "large"} onClick={onClick}>
```
same: `size` is already `"large"` by default as per `defaultProps`

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+        if (response.status == 404) {
+          setError("The survey you are trying to access does not exist. Please contact the administrator for further assistance.");
+        } else {
+          //setError("The survey could not be loaded at this time. Please try again later or contact the administrator for further assistance.");
```
what do we do with commented code, shall we uncomment it?

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+  // When a visit is loaded, record if it is locked
+  useEffect(() => {
+    setLocked(visit?.statusFlags && visit.statusFlags.includes("LOCKED"))
+  }, [visit])
```
```suggestion
  }, [visit]);
```

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+    let provider = getVisitField("provider");
+    provider = provider && provider.length > 1 ? provider.join(", ") : provider;
+    return (dateTime || location || provider) ?
+      <Alert variant="outlined" severity="info" icon={<EventNoteIcon/>} sx={{ marginTop: -2 }}>
```
`sx={{ mt: -2 }}`

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+        .map(s => {
+          let isComplete = isPageComplete(qId, s);
+          return (
+            <ListItem disablePadding sx={{ paddingRight: 6 }} secondaryAction={ isComplete && <DoneIcon color="success"/> } key={ s["@name"] } >
```
sx={{ pr: 6 }}

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Form.jsx:
```diff
@@ -0,0 +1,35 @@
+function Form(props) {
+
+  const actionSwitches = {
+    edit: data => (!["SUBMITTED", "LOCKED"].some(f => data.statusFlags.includes(f))),
```
optional: I'd add a comment `//Edit if form is neither SUBMITTED nor LOCKED`

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Patient.jsx:
```diff
@@ -0,0 +1,238 @@
+  const formatVisits = () => {
+    if (!visits) return;
+    setVisitGridRows(
+      visits?.map(visit => ({
```
no need for the optional chaining here, `visits` are defined at this point after check on line 129,
same for all lines below

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Patient.jsx:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  useEffect(fetchPatientData, []);
+  useEffect(fetchVisits, [patientData]);
```
why the hook depends on the `patientData` if `fetchVisits()` uniqly depends only on `patientUuid`?

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Patient.jsx:
```diff
@@ -0,0 +1,238 @@
+  },
+];
+
+function Patient(props) {
```
we don't use `props` anywhere

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SubjectView.jsx:
```diff
@@ -0,0 +1,54 @@
+      "label": "Created on",
+      "format": "date:yyyy-MM-dd HH:mm",
+    },
+  ]
```
Shall we also show `Created by`?

marta- replied:
No, it's irrelevant on the clinician dashboard.

acrowthe replied:
Excluding `created by` feels intentional - since the clinician dashboard disables creating new forms and subjects, the vast majority if not all forms/subjects will be created by backend service users.
Anything that isn't created by a service user will be created by an admin who have access to the regular dashboard which shows `created by`

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,94 @@
+        <CircularProgress size={24}/>
+      </IconButton>
+    </Tooltip>);
+  }
```
Let's space out blocks with one empty line

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+
+function Visit(props) {
+  const id = getSubjectIdFromPath(location.pathname);
+  const [ , patientUuid, visitUuid ] = /^([^/]+)\/([^/]+)$/.exec(id);
```
I suggest to guard against regex `.exec()` returning `null`

```
const match = /^([^/]+)\/([^/]+)$/.exec(id);
const patientUuid = match?.[1];
const visitUuid = match?.[2];
```

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+          setParents(json["parents"]);
+          setVisitInformation(json[VISIT_INFORMATION_FORM_TITLE]?.[0] || {});
+          let clinicPath = Object.values(json[VISIT_INFORMATION_FORM_TITLE]?.[0]).find(o => o?.question?.["@name"] == "clinic")?.value;
+          return fetchWithReLogin(globalLoginDisplay, `${clinicPath}.deep.json`)
```
```
         if (!clinicPath) {
            setError("Clinic is missing for this visit.");
            return;
          }
```

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+        ids.push(q);
+      }
+    });
+    !questionnaireIds && setQuestionnaireIds(ids);
```
Why do we need a separate state var for `questionnaireIds` if they are just `Object.keys(surveyData)`?

veronikaslc replied:
Morevover, we only reset `questionnaireIds` when there were none. When `questionnaireSetId` changes and we reload data, ids become stale forever.

marta- replied:
This looks like code adapted from QuestionnaireSet.jsx. There, we may have introduced questionnaireIds to be used as an effect dependency in an attempt to reduce the frequency of that effect in a different version of the code.

marta- replied:
I don't really mind this state var, definitely not a top refactoring priority.

marta- replied:
On that note, in a future iteration we should explore actually reusing, rather than adapting, code between clinician-dashboard/Visit.jsx and QuestionnaireSet.jsx, for example via a QuestionnaireSetUtils.

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+      .map(value => value.questionnaire['@name'])
+    setQuestionnaireSetIds(qids);
+
+    selectDataForQuestionnaireSet(visit, data, qids);
```
no need to explicitly include state var `visit` in this f-n call signature

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+    let ids = [];
+    let data = {};
+    questionnaireSetIds.forEach(q => {
+      if (visit[questionnaireSet?.[q]?.title]?.[0]?.['jcr:primaryType'] == "cards:Form") {
```
what if `visit` is `null`/`undefined`

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+
+  const displayVisitDateTime = () => {
+    let dateTimeAnswer = getVisitField("time");
+    return DateTime.fromISO(dateTimeAnswer).toLocaleString(DateTime.DATETIME_MED_WITH_WEEKDAY);
```
`getVisitField()` can return `null`, that makes `DateTime.fromISO(dateTimeAnswer)` invalid

```
  const dateTimeAnswer = getVisitField("time");
  if (!dateTimeAnswer) return null;

  const dt = DateTime.fromISO(dateTimeAnswer);
  return dt.isValid ? dt.toLocaleString(DateTime.DATETIME_MED_WITH_WEEKDAY) : null;
```

veronikaslc commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,409 @@
+      label={flag.substring(0,1).toUpperCase() + flag.substring(1).toLowerCase()}
+      variant="outlined"
+      size="small"
+      className={`${classes[flag + "Flag"] || classes.DefaultFlag}`}
```
`classes[flag + "Flag"] || classes.DefaultFlag` is a string already, the template literal `${ }` is unnecessary

acrowthe committed:
CARDS-2558/2609: Separate clinician dashboard and configurable user actions
- Address code review comments: Clean up and handle edge cases
- Clean up extensionUrl definition

marta- commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/SurveyLinkButton.jsx:
```diff
@@ -0,0 +1,100 @@
+    }, 5_000);
+  };
+
+  const fetchToken = () => {
+    setFetchingLink(true);
+    fetchWithReLogin(globalLoginDisplay, `${visitURL}.token.html`)
+      .then((response) => response.ok ? response.text() : Promise.reject(response))
+      .then((text) => {
+        let link = window.location.origin + "/Survey.html?auth_token=" + text.trim();
+        setSurveyLink(link);
+        copy(link);
+      })
+      .catch(() => setError("Could not generate survey link"))
+      .finally(() => setFetchingLink(false));
+  };
```
I would like to build in the ability to specify the validity of the token, or at least use a better default. For example, there could be an admin config that allows to specify how many days a token created this way would be valid from the moment of it's creation. This way, the functionality could be used to provide a new link to expired surveys in just one step. It doesn't have to be done in this PR unless it's trivial. Thoughts? @acrowthe @sdumitriu

acrowthe replied:
Implementing in #2220

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/Forms.jsx:
```diff
@@ -34,30 +34,30 @@ function Forms(props) {
-      "label": "Identifier",
-      "format": getEntityIdentifier,
-      "link": "dashboard+path",
+      key: "@name",
```
While I agree that it reads better without quotation marks around the keys, this kind of cosmetic change to a file that did not otherwise need updates for CARDS-2558 is not relevant for this PR and should have been made separately.

acrowthe replied:
Split into #2189

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -469,14 +473,24 @@ function Form (props) {
     );
   }
 
+  let isActionEnabled = (action) => (!!!actionSwitches || !!(actionSwitches[action]?.(data)));
+
+  let isDropdnEnabled = () => (
+    (isEdit && isActionEnabled("subject"))
+    || (!isEdit && isActionEnabled("text"))
+    || isActionEnabled("delete")
+  );
+
```
This would have been a really good time for implementing https://phenotips.atlassian.net/browse/CARDS-1084 . Maybe in the next iteration?

marta- commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Forms.jsx:
```diff
@@ -0,0 +1,65 @@
+    {
+      key: "",
+      label: "Subject",
+      format: (row) => (row.subject ? getHierarchy(row.subject, undefined, undefined, props.extensionURL) : ''),
```
Code style: why not extract extensionURL form props? This what we do in the majority of other components, and helps know which props are actually relevant.

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Update the module version

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Enable the clinician dashboard in test mode

acrowthe committed:
CARDS-2558: New "clinician-dashboard" module
Fix UI error when locking a visit

sdumitriu committed:
[misc] Refresh the yarn lockfile

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Add the new module to the list of modules to include in the standalone docker image

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Declare the proper dependencies to avoid code duplication

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Also copy the unsubscribe status in the patient info JSON

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Use copy-answers to access data about the visit

sdumitriu commented on modules/clinician-dashboard/src/main/frontend/src/clinician-dashboard/Visit.jsx:
```diff
@@ -0,0 +1,430 @@
+            disablePadding
+            secondaryAction={withAction && !isFormLocked(q) && <EditButton entryPath={surveyData?.[q]?.["@path"]}/>}
+          >
+            <ListItemButton onClick={() => navigate(`/content.html${surveyData?.[q]?.["@path"]}`)}>
```
It would be nice if these were actual links instead of just text with onclick handlers.

sdumitriu reviewed (commented):
Approved, but I'm the PR author.

sdumitriu committed:
CARDS-2558: New "clinician-dashboard" module
Fixed bad declaration of asset dependencies

acrowthe reviewed (approved):
Retested, looks good
